hack/generator/pkg/jsonast/swagger_type_extractor.go

```diff
@@ -93,6 +117,122 @@ func (extractor *typeExtractor) extractTypes(
 	return nil
 }
 
+// Look at the responses of the PUT to determine if this represents an ARM resource,
+// and if so, return the schema for it.
+// see: https://github.com/Azure/autorest/issues/1936#issuecomment-286928591
+func (extractor *SwaggerTypeExtractor) findARMResourceSchema(
+	filePath string,
+	swagger spec.Swagger,
+	op spec.Operation) *Schema {
+
+	if op.Responses != nil {
+		for statusCode, response := range op.Responses.StatusCodeResponses {
+			// only check OK and Created (per above linked comment)
+			if statusCode == 200 || statusCode == 201 {
+				result := extractor.getARMResourceSchemaFromResponse(filePath, swagger, response)
+				if result != nil {
+					return result
+				}
+			}
+		}
+	}
+
+	// none found
+	return nil
+}
+
+func (extractor *SwaggerTypeExtractor) getARMResourceSchemaFromResponse(filePath string, swagger spec.Swagger, response spec.Response) *Schema {
+	if response.Schema != nil {
+		// the schema can either be directly included
+
+		schema := MakeOpenAPISchema(
+			*response.Schema,
+			swagger,
+			filePath,
+			extractor.outputGroup,
+			extractor.outputVersion,
+			extractor.cache)
+
+		if isMarkedAsARMResource(schema) {
+			return &schema
+		}
+	} else if response.Ref.GetURL() != nil {
+		// or it can be under a $ref
+
+		filePath, swagger, refSchema := loadRefSchema(swagger, response.Ref, filePath, extractor.cache)
+		schema := MakeOpenAPISchema(
+			refSchema,
+			swagger,
+			filePath,
+			extractor.outputGroup,
+			extractor.outputVersion,
+			extractor.cache)
+
+		if isMarkedAsARMResource(schema) {
+			return &schema
+		}
+	}
+
+	return nil
+}
+
+// a Schema represents an ARM resource if it (or anything reachable via $ref or AllOf)
+// is marked with x-ms-azure-resource, or if it (or anything reachable via $ref or AllOf)
+// has each of the properties: id,name,type.
+func isMarkedAsARMResource(schema Schema) bool {
+	hasID := false
+	hasName := false
+	hasType := false
+
+	var recurse func(schema Schema) bool
```

Comment:
It's weird that this is necessary - I don't understand why they couldn't have made recurse visible inside the function if it was defined with `:=`. :shrug: